File: plugin/scala/src/main/scala/com/twosigma/beaker/scala/util/SparkContextManager.scala
/*
 *  Copyright 2014-2016 TWO SIGMA OPEN SOURCE, LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
 
package com.twosigma.beaker.scala.util

import java.util.concurrent.ArrayBlockingQueue
import scala.tools.nsc.Settings
import scala.tools.nsc.interpreter.Results.Error
import scala.tools.nsc.interpreter.Results.Success
import scala.tools._ 
import nsc.interpreter.{Completion, CompletionAware, IMain, 
JLineCompletion, JLineDelimiter, JList, Parsed, Results, IR} 
import Completion.{Candidates, ScalaCompleter} 

import jline.console.completer.{Completer, ArgumentCompleter} 
import java.util.ArrayList
import java.util.Map
import java.io.PrintStream
import com.twosigma.beaker.jvm.`object`.SparkProgressService
import scala.collection.JavaConversions._
import scala.util.Random

import org.apache.spark._
import org.apache.spark.rdd.RDD
import org.apache.spark.scheduler._

import com.twosigma.base.log4j.TSLogger

class SparkContextManager {
  private val random = new Random()
  private val log = TSLogger.getTSLogger()

  private var running = false
  private var error = ""
  private var sc: SparkContext = null

  private var conf = new SparkConf().setAppName("beaker")
    .setMaster("spark://localhost")
    .set("spark.cores.max", "10")
    .set("spark.executor.memory", "8g")

  def configure(configuration: java.util.Map[String, String]) = {
    if (configuration.containsKey("master")) {
      val master = configuration.get("master")
      log.debug(s"The master is $master.")
      conf.setMaster(master)
    }
    configuration.foreach {
      case ("master", _) => ()
      case (key, value) => {
        conf.set(key, value)
        log.debug(s"Setting $key = $value")
      }
    }
  }

  def start(progressService: SparkProgressService): SparkContext = {
    log.trace("Initializing SparkContext")
    error = ""

    if (!running || sc == null) {
      try {
        // assign new random port for Spark UI every time the context is (re)created
        conf.set("spark.ui.port", (random.nextInt(4080 - 4040) + 4040).toString())
        sc = new SparkContext(conf)
        sc.addSparkListener(new SparkListener() {
          override def onApplicationStart(applicationStart: SparkListenerApplicationStart) {
            //log.debug("Spark ApplicationStart: " + applicationStart.appName)
            progressService.applicationStart(applicationStart.appName)
          }
          override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd) {
            //log.debug("Spark ApplicationEnd: " + applicationEnd.time)
            progressService.applicationEnd()
          }
          override def onJobStart(jobStart: SparkListenerJobStart) {
            //log.debug("Spark JobStart: " + jobStart.jobId)
            progressService.jobStart(jobStart.jobId)
          }
          override def onJobEnd(jobEnd: SparkListenerJobEnd) {
            //log.debug("Spark JobEnd: " + jobEnd.time)
            progressService.jobEnd(jobEnd.jobId)
          }
          override def onStageSubmitted(stageSubmitted: SparkListenerStageSubmitted) {
            //log.debug(s"Spark StageSubmitted: ${stageSubmitted.stageInfo.stageId} [${stageSubmitted.stageInfo.name}] (${stageSubmitted.stageInfo.numTasks} tasks started)")
            progressService.stageStart(stageSubmitted.stageInfo.stageId, stageSubmitted.stageInfo.numTasks)
          }
          override def onStageCompleted(stageCompleted: SparkListenerStageCompleted) {
            //log.debug(s"Spark StageCompleted: ${stageCompleted.stageInfo.stageId} [${stageCompleted.stageInfo.name}]  (${stageCompleted.stageInfo.numTasks} tasks started)")
            progressService.stageEnd(stageCompleted.stageInfo.stageId, stageCompleted.stageInfo.failureReason.getOrElse(null))
          }
          override def onTaskStart(taskStart: SparkListenerTaskStart) {
            //log.debug(s"Spark TaskStart: ${taskStart.taskInfo.taskId} (Stage ${taskStart.stageId})")
            progressService.taskStart(taskStart.stageId, taskStart.taskInfo.taskId)
          }
          override def onTaskEnd(taskEnd: SparkListenerTaskEnd) {
            //log.debug(s"Spark TaskEnd: ${taskEnd.taskInfo.taskId} (Stage ${taskEnd.stageId})")
            progressService.taskEnd(taskEnd.stageId, taskEnd.taskInfo.taskId, taskEnd.taskInfo.failed)
          }
        })
      } catch {
        case e: java.net.BindException => {
          log.error(s"BindException during SparkContext initialization: ${e.getMessage}")
        }
        case e: Throwable => {
          log.error(s"Error during SparkContext initialization: ${e.getMessage}")
        }
      }
    }
    running = true
    sc
  }

  def stop = {
    try {
      if (sc != null) {
        sc.stop()
      }
      running = false
    } catch {
      case e: Throwable => {
        log.error(s"Error while stopping SparkContext: ${e.getMessage}")
      }
    }
  }

  def getContext: SparkContext = {
    sc
  }

  def getConfiguration: SparkConf = {
    conf
  }

  def isRunning: Boolean = {
    running
  }

  def cancelAllJobs() = {
    sc.cancelAllJobs()
  }

  def getError: String = {
    error
  }
}

object SparkContextManagerObj extends SparkContextManager {}

File: kernel/sparkex/src/main/java/com/twosigma/beakerx/widget/SparkContextManager.java
/*
 *  Copyright 2018 TWO SIGMA OPEN SOURCE, LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.twosigma.beakerx.widget;

import com.twosigma.beakerx.TryResult;
import com.twosigma.beakerx.jvm.object.SimpleEvaluationObject;
import com.twosigma.beakerx.kernel.KernelFunctionality;
import com.twosigma.beakerx.kernel.KernelManager;
import com.twosigma.beakerx.message.Message;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.scheduler.SparkListener;
import org.apache.spark.scheduler.SparkListenerJobEnd;
import org.apache.spark.scheduler.SparkListenerJobStart;
import org.apache.spark.scheduler.SparkListenerStageCompleted;
import org.apache.spark.scheduler.SparkListenerStageSubmitted;
import org.apache.spark.scheduler.SparkListenerTaskEnd;
import org.apache.spark.scheduler.SparkListenerTaskStart;
import org.apache.spark.sql.SparkSession;
import scala.Tuple2;
import scala.collection.Iterator;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import static com.twosigma.beakerx.kernel.PlainCode.createSimpleEvaluationObject;

public class SparkContextManager {

  public static final String SPARK_MASTER = "spark.master";
  public static final String SPARK_EXECUTOR_MEMORY = "spark.executor.memory";
  public static final String SPARK_APP_NAME = "spark.app.name";
  public static final String SPARK_CORES_MAX = "spark.cores.max";
  public static final String SPARK_EXECUTOR_CORES = "spark.executor.cores";

  private final SparkUI sparkUI;
  private Map<Integer, SparkStateProgress> progressBars = new HashMap<>();
  private VBox jobPanel = null;
  private HBox statusPanel;
  private VBox sparkView;
  private Text masterURL;
  private Text executorMemory;
  private Text sparkContextAlias;
  private Text sparkSessionAlias;
  private Text executorCores;

  private SparkSession.Builder sparkSessionBuilder;

  public SparkContextManager(SparkUI sparkUI, SparkSession.Builder sparkSessionBuilder) {
    this.sparkUI = sparkUI;
    this.sparkSessionBuilder = sparkSessionBuilder;
    SparkVariable.putSparkContextManager(getSparkConf(), this);
    createSparkView();
  }

  private SparkConf createSparkConf() {
    SparkConf sparkConf = new SparkConf();
    try {
      Field options = this.sparkSessionBuilder.getClass().getDeclaredField("org$apache$spark$sql$SparkSession$Builder$$options");
      options.setAccessible(true);
      Iterator iterator = ((scala.collection.mutable.HashMap) options.get(this.sparkSessionBuilder)).iterator();
      while (iterator.hasNext()) {
        Tuple2 x = (Tuple2) iterator.next();
        sparkConf.set((String) (x)._1, (String) (x)._2);
      }
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
    return sparkConf;
  }

  private SparkConf getSparkConf() {
    return createSparkConf();
  }

  private SparkContext sparkContext() {
    return sparkSessionBuilder.getOrCreate().sparkContext();
  }

  private void createSparkView() {
    this.masterURL = createMasterURL();
    this.executorMemory = createExecutorMemory();
    this.executorCores = createExecutorCores();
    this.sparkContextAlias = sparkContextAlias();
    this.sparkSessionAlias = sparkSessionAlias();
    Button connect = createConnectButton();
    ArrayList<Widget> children = new ArrayList<>();
    children.add(masterURL);
    children.add(executorCores);
    children.add(executorMemory);
    children.add(sparkContextAlias);
    children.add(sparkSessionAlias);
    children.add(connect);
    VBox vBox = new VBox(children);
    this.sparkUI.add(vBox);
    this.sparkView = vBox;
  }

  private Text createExecutorCores() {
    Text cores = new Text();
    cores.setDescription("Executor cores");
    if (getSparkConf().contains(SPARK_EXECUTOR_CORES)) {
      cores.setValue(getSparkConf().get(SPARK_EXECUTOR_CORES));
    } else {
      cores.setValue("10");
    }
    return cores;
  }

  private Text sparkContextAlias() {
    Text alias = new Text();
    alias.setDescription("SparkContext alias");
    alias.setValue("sc");
    return alias;
  }

  private Text sparkSessionAlias() {
    Text alias = new Text();
    alias.setDescription("SparkSession alias");
    alias.setValue("spark");
    return alias;
  }

  private Text createExecutorMemory() {
    Text masterURL = new Text();
    masterURL.setDescription("Executor Memory");
    if (getSparkConf().contains(SPARK_EXECUTOR_MEMORY)) {
      masterURL.setValue(getSparkConf().get(SPARK_EXECUTOR_MEMORY));
    } else {
      masterURL.setValue("8g");
    }
    return masterURL;
  }

  private Text createMasterURL() {
    Text masterURL = new Text();
    masterURL.setDescription("Master URL");
    if (getSparkConf().contains(SPARK_MASTER)) {
      masterURL.setValue(getSparkConf().get(SPARK_MASTER));
    }
    return masterURL;
  }

  private Button createConnectButton() {
    Button connect = new Button();
    connect.setDescription("Connect");
    connect.registerOnClick((content, message) -> initSparkContext(message));
    return connect;
  }

  private void initSparkContext(Message parentMessage) {
    KernelFunctionality kernel = KernelManager.get();
    try {
      SparkConf sparkConf = configureSparkConf(getSparkConf());
      sparkSessionBuilder.config(sparkConf);
      SparkSession sparkSession = getSparkSession();
      addListener(sparkContext());
      SparkVariable.putSparkContext(sparkContext());
      SparkVariable.putSparkSession(sparkSession);
      TryResult tryResultSparkContext = initSparkContextInShell(kernel);
      if (tryResultSparkContext.isError()) {
        sendError(parentMessage, kernel, tryResultSparkContext.error());
      }
    } catch (Exception e) {
      sendError(parentMessage, kernel, e.getMessage());
    }
  }

  private SparkSession getSparkSession() {
    return sparkSessionBuilder.getOrCreate();
  }

  private TryResult initSparkContextInShell(KernelFunctionality kernel) {
    if (sparkContextAlias.getValue() == null || sparkContextAlias.getValue().isEmpty()) {
      throw new RuntimeException("SparkContext alias can not be empty");
    }
    if (sparkSessionAlias.getValue() == null || sparkSessionAlias.getValue().isEmpty()) {
      throw new RuntimeException("SparkContext alias can not be empty");
    }
    String addSc = String.format(
            "import com.twosigma.beakerx.widget.SparkVariable\n" +
                    "var %s = SparkVariable.getSparkContext()\n" +
                    "var %s = SparkVariable.getSparkSession()\n",
            sparkContextAlias.getValue(), sparkSessionAlias.getValue());

    SimpleEvaluationObject seo = createSimpleEvaluationObject(addSc, kernel, new Message(), 1);
    return kernel.executeCode(addSc, seo);
  }

  private void sendError(Message parentMessage, KernelFunctionality kernel, String message) {
    SimpleEvaluationObject seo = createSimpleEvaluationObject("", kernel, parentMessage, 1);
    seo.error(message);
  }

  private SparkConf configureSparkConf(SparkConf sparkConf) {
    if (!sparkConf.contains(SPARK_APP_NAME)) {
      sparkConf.setAppName("beaker_" + UUID.randomUUID().toString());
    }
    if (masterURL.getValue() != null && !masterURL.getValue().isEmpty()) {
      sparkConf.set(SPARK_MASTER, masterURL.getValue());
    }
    if (!isLocalSpark(sparkConf)) {
      sparkConf.set("spark.repl.class.outputDir", KernelManager.get().getOutDir());
    }
    if (executorMemory.getValue() != null && !executorMemory.getValue().isEmpty()) {
      sparkConf.set(SPARK_EXECUTOR_MEMORY, executorMemory.getValue());
    }

    if (executorCores.getValue() != null && !executorCores.getValue().isEmpty()) {
      sparkConf.set(SPARK_EXECUTOR_CORES, executorCores.getValue());
    }

    if (!sparkConf.contains(SPARK_CORES_MAX)) {
      sparkConf.set(SPARK_CORES_MAX, "100");
    }
    return sparkConf;
  }

  private static boolean isLocalSpark(SparkConf sparkConf) {
    return sparkConf.contains(SPARK_MASTER) && sparkConf.get(SPARK_MASTER) != null && sparkConf.get("spark.master").startsWith("local");
  }

  private SparkContext addListener(SparkContext sc) {
    sc.addSparkListener(new SparkListener() {

      @Override
      public void onJobStart(SparkListenerJobStart jobStart) {
        super.onJobStart(jobStart);
      }

      @Override
      public void onJobEnd(SparkListenerJobEnd jobEnd) {
        super.onJobEnd(jobEnd);
      }

      @Override
      public void onStageSubmitted(SparkListenerStageSubmitted stageSubmitted) {
        super.onStageSubmitted(stageSubmitted);
        startStage(stageSubmitted.stageInfo().stageId(), stageSubmitted.stageInfo().numTasks());
      }

      @Override
      public void onStageCompleted(SparkListenerStageCompleted stageCompleted) {
        super.onStageCompleted(stageCompleted);
        endStage(stageCompleted.stageInfo().stageId());
      }

      @Override
      public void onTaskStart(SparkListenerTaskStart taskStart) {
        super.onTaskStart(taskStart);
        taskStart(taskStart.stageId(), taskStart.taskInfo().taskId());
      }

      @Override
      public void onTaskEnd(SparkListenerTaskEnd taskEnd) {
        super.onTaskEnd(taskEnd);
        taskEnd(taskEnd.stageId(), taskEnd.taskInfo().taskId());
      }
    });
    return sc;
  }

  public void applicationStart() {
    sparkUI.removeDOMWidget(sparkView);
    sparkView = null;
    statusPanel = createStatusPanel();
  }

  public void applicationEnd() {
    if (statusPanel != null) {
      sparkUI.removeDOMWidget(statusPanel);
      statusPanel = null;
      createSparkView();
    }
  }

  private HBox createStatusPanel() {
    Label appStatus = createAppStatus();
    Button disconnect = createDisconnectButton();
    HBox statusPanel = new HBox(Arrays.asList(uiLink(), disconnect, appStatus));
    sparkUI.add(statusPanel);
    return statusPanel;
  }

  private Label createAppStatus() {
    Label appStatus = new Label();
    appStatus.setValue("Connected to " + getSparkConf().get("spark.master"));
    return appStatus;
  }

  private Button createDisconnectButton() {
    Button disconnect = new Button();
    disconnect.registerOnClick((content, message) -> sparkContext().stop());
    disconnect.setDescription("Disconnect");
    return disconnect;
  }

  private void startStage(int stageId, int numTasks) {
    SparkStateProgress intProgress = new SparkStateProgress(numTasks, stageId, stageId, jobLink(stageId), stageLink(stageId));
    intProgress.init();
    if (jobPanel != null) {
      jobPanel.getLayout().setDisplayNone();
      jobPanel.close();
    }
    jobPanel = new VBox(new ArrayList<>());
    jobPanel.add(intProgress);
    jobPanel.display();
    progressBars.put(stageId, intProgress);
  }

  private void endStage(int stageId) {
    SparkStateProgress sparkStateProgress = progressBars.get(stageId);
    sparkStateProgress.hide();
  }

  private void taskStart(int stageId, long taskId) {
    SparkStateProgress intProgress = progressBars.get(stageId);
    intProgress.addActive();
  }

  private void taskEnd(int stageId, long taskId) {
    SparkStateProgress intProgress = progressBars.get(stageId);
    intProgress.addDone();
  }

  private HTML uiLink() {
    if (sparkContext().uiWebUrl().isDefined()) {
      HTML html = new HTML();
      html.setValue("<a target=\"_blank\" href=\"" + sparkContext().uiWebUrl().get() + "\">Spark UI" + "</a>");
      return html;
    } else {
      HTML html = new HTML();
      html.setValue("<a target=\"_blank\" href=\"\">Spark UI " + "</a>");
      return html;
    }
  }

  private String stageLink(int stageId) {
    if (sparkContext().uiWebUrl().isDefined()) {
      return sparkContext().uiWebUrl().get() + "/stages/stage/?id=" + stageId + "&attempt=0";
    } else {
      return "";
    }
  }

  private String jobLink(int jobId) {
    if (sparkContext().uiWebUrl().isDefined()) {
      return sparkContext().uiWebUrl().get() + "/jobs/job/?id=" + jobId;
    } else {
      return "";
    }
  }

  public void cancelAllJobs() {
    if (sparkContext() != null) {
      sparkContext().cancelAllJobs();
    }
  }
}
